app/models/user.rb

```diff
@@ -49,13 +51,26 @@ class User < ActiveRecord::Base
   def self.authenticate(login, password)
     user = find_by_login(login.downcase) || find_by_email(login.downcase) # need to get the salt
 
-    if user and user.password == encrypt_password(password)
-      return user
+    if user
+      if user[:password].length != 0
+        if user.password_before_type_cast == encrypt_password(password)
+          user[:password] = ""
+          user[:encrypted_password] = devise_encrypt_password(password)
+          user.save
+          return user
+        end
+      elsif user.valid_password?(password)
+        return user
+      end
```

Comment:
Can we have specs for this logic? It is quite complicated and I think it will be worth it to have specs for the different scenarios that can happen...